ggml : move rope type enum to ggml.h

danbev opened the pull request:
This commit moves the `llama_rope_type` enum from `llama.h` to `ggml.h` and changes its name to `ggml_rope_type`.

The motivation for this change is to address the TODO in `llama.h` and use the enum in ggml.

Note: This commit does not change the `mode` parameter to be of type `enum ggml_rope_type`. The name `mode` and its usage suggest that it might be more generic and possibly used as a bit field for multiple flags. Further investigation/discussion may be needed to determine if `mode` should be restricted to RoPE types.



- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

danbev committed:
ggml : move rope type enum to ggml.h

This commit moves the `llama_rope_type` enum from `llama.h` to
`ggml.h` and changes its name to `ggml_rope_type`.

The motivation for this change is to address the TODO in `llama.h` and
use the enum in ggml.

Note: This commit does not change the `mode` parameter to be of type
`enum ggml_rope_type`. The name `mode` and its usage suggest that it
might be more generic and possibly used as a bit field for multiple
flags. Further investigation/discussion may be needed to determine
if `mode` should be restricted to RoPE types.

ggerganov commented on include/llama.h:
```diff
@@ -462,7 +453,7 @@ extern "C" {
 
     LLAMA_API enum llama_vocab_type   llama_vocab_type  (const struct llama_model * model);
-    LLAMA_API enum llama_rope_type    llama_rope_type   (const struct llama_model * model);
+    LLAMA_API enum ggml_rope_type     ggml_rope_type    (const struct llama_model * model);
```
This should continue to be called `llama_rope_type` since it is part of the llama API:

```suggestion
    LLAMA_API enum ggml_rope_type     llama_rope_type   (const struct llama_model * model);
```

danbev replied:
I've updated this now. Thanks

slaren commented:
If it is a bit field, then my preference would be to declare the different values as constants rather than as an enum. However, I don't think it is a bit field at the moment, even if it used to be in the past. Also, support for GLM rope was removed a while ago, so from what I can tell, the only two valid options are either normal or neox. `GGML_ROPE_TYPE_NONE` also doesn't really make sense in ggml, I understand that the reason it exists in llama.cpp is to support models that don't use rope, but that shouldn't be carried to ggml.

ggerganov commented:
Yes make sense. Probably we need `ggml_rope_type` with "norm" and "neox" in `ggml` and then another enum `llama_rope_type` in `llama` that combines the `ggml_rope_type` values with an extra "skip" value.

danbev committed:
squash! ggml : move rope type enum to ggml.h

This commit removes GGML_ROPE_TYPE_NONE and GGML_ROPE_TYPE_GLM from
ggml.h, and back the llama_rope_type enum.

I've kept the assert for GGML_ROPE_TYPE_GLM as I'm not sure if it is
safe to remove it yet.

danbev commented:
@slaren @ggerganov Thanks for the reviews/feedback! I've updated with a commit containing your suggestions as I understood them.

Would it makes sense to follow up this PR with one that updates the `mode` parameter to be of type `ggml_rope_type` for the `ggml_rope_*` functions?

ggerganov commented:
> If it is a bit field, then my preference would be to declare the different values as constants rather than as an enum.

@slaren Sorry, I overlooked this yesterday. What's the advantage of having the values as constants compared to enum? And to become a bit field, do you mean that the `0` values should be changed to `1`?

@danbev It would be nice to make the change in a single PR. But let's first clarify how to express the rope types before we proceed.

ggerganov commented:
Ok, got it. So how about we define NeoX type like this and we don't even need the "norm" type within `ggml`:

```diff
diff --git a/ggml/include/ggml.h b/ggml/include/ggml.h
index 22f5dfed..a7ee17d2 100644
--- a/ggml/include/ggml.h
+++ b/ggml/include/ggml.h
@@ -244,6 +244,8 @@
 #define GGML_EXIT_SUCCESS 0
 #define GGML_EXIT_ABORTED 1
 
+#define GGML_ROPE_TYPE_NEOX 2
+
 #define GGUF_MAGIC "GGUF"
 
 #define GGUF_VERSION 3
@@ -437,12 +439,6 @@ extern "C" {
         GGML_FTYPE_MOSTLY_Q4_0_8_8 = 27, // except 1d tensors
     };
 
-    // Rotary Positional Embedding (RoPE) types
-    enum ggml_rope_type {
-        GGML_ROPE_TYPE_NORM =  0,
-        GGML_ROPE_TYPE_NEOX =  2,
-    };
-
     // available tensor operations:
     enum ggml_op {
         GGML_OP_NONE = 0,
```

And we keep the `mode` argument `ggml_rope` as `int` so that in the future we could use it as a bit field to store multiple options if needed.

@danbev We would need to update `mode & 2` -> `mode & GGML_ROPE_TYPE_NEOX` in the backends as ell.

danbev committed:
squash! ggml : move rope type enum to ggml.h

This commit removes the enum ggml_rope_type from ggml.h and replaces it
with a define (GGML_ROPE_TYPE_NEOX). This define is used in the code to
check if the mode is set to GPT-NeoX. Also the enum llama_rope_type has
been updated to reflect this change.

danbev committed:
squash! ggml : move rope type enum to ggml.h

This commit contains a suggestion enable the GGML_ROPE_TYPE_NEOX
macro/define to be passed to the shader compiler.

danbev committed:
squash! ggml : move rope type enum to ggml.h

This commit fixes the editorconfig-checker warnings.

ggerganov commented on ggml/include/ggml.h:
```diff
@@ -1454,7 +1456,7 @@ extern "C" {
 
     // rotary position embedding
     // if mode & 1 == 1, skip n_past elements (NOT SUPPORTED)
-    // if mode & 2 == 1, GPT-NeoX style
```
Update the comment to:

```
    // if (mode & 1) - skip n_past elements (NOT SUPPORTED)
    // if (mode & GGML_ROPE_TYPE_NEOX) - GPT-NeoX style
```

danbev replied:
I've updated this now :+1:

ggerganov commented on ggml/src/CMakeLists.txt:
```diff
@@ -671,7 +685,10 @@ if (GGML_KOMPUTE)
+                ${GGML_HEADER_PATH}
+                COMMAND ${glslc_executable} --target-env=vulkan1.2
+                        -o ${spv_file} ${CMAKE_CURRENT_SOURCE_DIR}/${source}
+                        -DGGML_ROPE_TYPE_NEOX=${GGML_ROPE_TYPE_NEOX_VALUE}
```
No need for the kompute-related changes in the CMake build system. Simply define the `GGML_ROPE_TYPE_NEOX` constant in the kompute shaders where needed for now

danbev replied:
Ah great, I've added this to `rope_common.comp` in https://github.com/ggerganov/llama.cpp/pull/8949/commits/cfb98641828d79d7ffba94f74b970aed133a520b.

danbev committed:
squash! ggml : move rope type enum to ggml.h

Update comment for ggml_rope function.

danbev committed:
Revert "squash! ggml : move rope type enum to ggml.h"

This reverts commit 6261222bd0dc0efd51f0fb0435ad3f16a5b52fd6.

danbev committed:
squash! ggml : move rope type enum to ggml.h

Add GGML_ROPE_TYPE_NEOX to rope_common.comp.

slaren commented on include/llama.h:
```diff
@@ -95,15 +95,13 @@ extern "C" {
+        LLAMA_ROPE_TYPE_NEOX = GGML_ROPE_TYPE_NEOX,
     };
 
+
```
```suggestion
```

slaren committed:
remove extra line
